### PR TITLE
[hop][dynamo] support torch.SymInt inputs

### PR DESCRIPTION
**Pull request by ydwu4**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141524
* #142185

Fixes https://github.com/pytorch/pytorch/issues/141305.

```python
        class M(torch.nn.Module):
            def forward(self, x, y, z):
                a = y.shape[0]
                b = z.shape[0]

                def true_fn(x):
                    return x + a

                def false_fn(x):
                    return x + b * z

                # When exporting with non-strict: a and b are symints,
                # so torch.compile need to wrap and trace symint inputs.
                return torch.cond(x.shape[0] > 5, true_fn, false_fn, (x,))
```

In non-strict export, when inputs are annotated with dynamic shape, the a, and b in above example are torch.SymInt type. true_fn and false_fn will have closure that're of torch.SymInt types.  The error is triggered because we didn't handle SymInt inputs in dynamo and ends up using a UserDefinedObjectVariable for it, which doesn't have a proxy. We added support by following how we handle SymBool input previously.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

**Comment by ydwu4** [comment]
Consider the following user code, where x.shape[0] is dynamic:
```python
def f(x):
  s0 = x.shape[0]

  def true_fn(x):
     if isinstance(s0, int):
       return x + s0
     else:
       return x - s0

  torch.cond(s0 > 4, true_fn, ...)
```

When we run this code eagerly, s0 is an integer, torch.cond sees an integer closure. But when we non-strict exporting this code s0 is a SymInt. From torch.cond's point view, the `s0` closure is created from a `symint` input, so isinstance(s0, int) returns False. However, this is not what we want: the graph we get from tracing the compiled function has different semantic  as eager code.

We could say that isinstance(s0, int) should always return True no matter the SymNodeVariable s0 is created from an integer or a symint (i.e. SymNodeVariable.python_type() should always return int). This makes the above code produce correct graph. However, there is a test case  test_default_partitioner_output_tensor_shape_tensor that relies on a SymInt input's python type being SymInt (i.e. SymNodeVariable.python_type() should return the input type).

So we have two options here:
1. either we make SymNodeVariable.python_type  to be always int (non-strict export will be happy)
2. or we make SymNodeVariable.python_type to return symint or int depending on the input type (it's more intuitive.)

Any suggestions? @zou3519

In fact, I guess this is a limitation of non-strict mode I think:
```
    def test_nonstrict_isinstance(self):
        from torch.export import Dim

        class M(torch.nn.Module):
            def forward(self, x, y, z):
                a = y.shape[0]
                if isinstance(a, int):
                    return x + y.sum() + z
                else:
                    return x - y.sum() - z

        args = (torch.ones(3, 3), torch.ones(5, dtype=torch.int32), torch.ones(3, 3))
        model = M()
        dynamic_shapes = {"x": {0: Dim("d")}, "y": {0: Dim("d1")}, "z": {0: Dim("d")}}
        torch.export.export(model, args, dynamic_shapes=dynamic_shapes, strict=False)
```

Where it produces a graph:
```
ExportedProgram:
    class GraphModule(torch.nn.Module):
        def forward(self, x: "f32[s0, 3]", y: "i32[s1]", z: "f32[s0, 3]"):
             # File: /data/users/yidi/pytorch/test/functorch/test_control_flow.py:6270 in forward, code: return x - y.sum() - z
            sum_1: "i64[]" = torch.ops.aten.sum.default(y);  y = None
            sub: "f32[s0, 3]" = torch.ops.aten.sub.Tensor(x, sum_1);  x = sum_1 = None
            sub_1: "f32[s0, 3]" = torch.ops.aten.sub.Tensor(sub, z);  sub = z = None
            return (sub_1,)
```

File: torch/_dynamo/variables/builder.py [hunk 1]
```diff
@@ -994,6 +994,48 @@ def build_key_value(i, k, v):
                 sym_node_proxy,
                 new_symint == 1,
             )
+        elif isinstance(value, torch.SymInt):
```
Comment (zou3519) [inline torch/_dynamo/variables/builder.py:997]: Is it possible to deduplicate this with the torch.SymBool branch?

**Review by zou3519 — COMMENTED** [review]
Seems fine to me, I just want to know if we can deduplicate

**Comment by zou3519** [comment]
> So we have two options here:

Got some advice from Ed, which sounds like the following:
1) We should try to treat SymInts as not-user-facing. That is, any SymInts we see are actually user ints, so isinstance(s0, int) should return True.
2) The way to justify this for the cond situation is: If there is a SymInt in non-strict export that we are passing to torch.cond, then this SymInt should be representing a user int. So it is OK for torch.compile (in torch.cond) to treat this SymInt like an integer. (with isinstance(s0, int) == True)
3) On testing: we should really stop running torch.compile over tests that use raw SymInts. Let's blanket ban dynamo-wrapped tests over functorch/test_aotdispatch.py (I thought I did that already, but I guess not).

**Comment by ezyang** [comment]
I'm not going to review the code unless explicitly asked to

File: torch/_dynamo/variables/builder.py [hunk 2]
```diff
@@ -944,40 +944,53 @@ def build_key_value(i, k, v):
...
+                    # We need to create an unbacked symint to replace the unbacked symbool.
+                    new_symint = self.tx.output.shape_env.create_unbacked_symint()
+                else:
+                    # TODO (yidi): we need to figure out a way to propagate the guards
+                    # we accumulated when tracing the subggraph to outer shape_env. For normal symints,
+                    # this is automatically done by evaluating the guards once but this
+                    # will cause data-dependent error when we evaluate the outer unbacked symints.
+                    # The test case that triggers this graph break is test_cond_unbacked_symint_closure
+                    unimplemented(
+                        "unbacked symint input is not supported yet. If you need this feature, please file a github issue."
```
Comment (zou3519) [inline torch/_dynamo/variables/builder.py:982]: nit: "unbacked symints as inputs to ..."

File: torch/_dynamo/variables/builder.py [hunk 3]
```diff
@@ -944,40 +944,53 @@ def build_key_value(i, k, v):
...
-                        new_source,
+                        int(value.node.hint),
+                        source,
                         dynamic_dim=DimDynamic.DYNAMIC,
                     )
                 )
             else:
-                # We need to create an unbacked symint to replace the unbacked symbool.
-                new_symint = self.tx.output.shape_env.create_unbacked_symint()
+                if isinstance(value, torch.SymBool):
```
Comment (zou3519) [inline torch/_dynamo/variables/builder.py:972]: I expected some logic for "we're only going to handle symint/symbool inputs to torch.compile if we're in non-strict export", what happened to that?

Comment (ydwu4) [reply torch/_dynamo/variables/builder.py:972]: We have quite some tests that does make_fx(symbolic=True) and then the shape expressions triggers the check here.

**Review by zou3519 — APPROVED** [review]

File: torch/_dynamo/variables/builder.py (hunk 3 above)
Comment (zou3519) [reply torch/_dynamo/variables/builder.py:972]: yike

**Pull request closed without merging**